File: packages/sdk-csharp/Melaya/AuthApi.cs
using System.Text.Json;

namespace Melaya;

/// <summary>
/// Auth API — login, register, MFA, password management, and session utilities.
/// Maps to <c>/api/v1/private/auth/*</c> (authenticated) and public counterparts.
/// </summary>
public sealed class AuthApi
{
    private readonly MelayaHttpClient _http;

    internal AuthApi(MelayaHttpClient http) => _http = http;

    // ── Public endpoints ──────────────────────────────────────────────────────

    /// <summary>
    /// Password + username login. Returns a session JWT and optional MFA challenge token.
    /// On MFA-enabled accounts, use the returned <c>ChallengeToken</c> with <see cref="VerifyMfaAsync"/>.
    /// </summary>
    public async Task<LoginResult> LoginAsync(string username, string password, CancellationToken ct = default)
    {
        var body = new { username, password };
        return await _http.PostAsync<LoginResult>("/api/v1/private/auth/login", body, ct).ConfigureAwait(false);
    }

    /// <summary>
    /// Resolve an MFA challenge after login. Exchange the challenge token + TOTP/recovery code
    /// for a full session JWT.
    /// </summary>
    public async Task<LoginResult> VerifyMfaAsync(string challengeToken, string code, CancellationToken ct = default)
    {
        var body = new { challengeToken, code };
        return await _http.PostAsync<LoginResult>("/api/v1/private/auth/mfa/verify", body, ct).ConfigureAwait(false);
    }

    /// <summary>Create a new user account. Sends email verification.</summary>
    public async Task<BoolResult> RegisterAsync(string username, string email, string password, CancellationToken ct = default)
    {
        var body = new { username, email, password };
        return await _http.PostAsync<BoolResult>("/api/v1/private/auth/register", body, ct).ConfigureAwait(false);
    }

    /// <summary>Confirm email address from signup link token.</summary>
    public async Task<BoolResult> VerifySignupAsync(string token, CancellationToken ct = default)
    {
        var body = new { token };
        return await _http.PostAsync<BoolResult>("/api/v1/private/auth/verify-signup", body, ct).ConfigureAwait(false);
    }

    /// <summary>Re-send email verification message.</summary>
    public async Task<BoolResult> ResendVerificationAsync(string email, CancellationToken ct = default)
    {
        var body = new { email };
        return await _http.PostAsync<BoolResult>("/api/v1/private/auth/resend-verification", body, ct).ConfigureAwait(false);
    }

    /// <summary>Initiate password reset flow; sends email with reset token.</summary>
    public async Task<BoolResult> ForgotPasswordAsync(string email, CancellationToken ct = default)
    {
        var body = new { email };
        return await _http.PostAsync<BoolResult>("/api/v1/private/auth/forgot-password", body, ct).ConfigureAwait(false);
    }

    /// <summary>Complete password reset using token from email.</summary>
    public async Task<BoolResult> ResetPasswordAsync(string token, string newPassword, CancellationToken ct = default)
    {
        var body = new { token, newPassword };
        return await _http.PostAsync<BoolResult>("/api/v1/private/auth/reset-password", body, ct).ConfigureAwait(false);
    }

    // ── Authenticated endpoints ───────────────────────────────────────────────

    /// <summary>Return current authenticated user profile (id, username, email, tier, role, capabilities).</summary>
    public async Task<UserProfile> MeAsync(CancellationToken ct = default)
    {
        return await _http.GetAsync<UserProfile>("/api/v1/private/auth/me", ct: ct).ConfigureAwait(false);
    }

    /// <summary>Lightweight session validity check; returns <c>ok: true</c> if the token is valid.</summary>
    public async Task<BoolResult> CheckAsync(CancellationToken ct = default)
    {
        return await _http.GetAsync<BoolResult>("/api/v1/private/auth/check", ct: ct).ConfigureAwait(false);
    }

    /// <summary>Change password (requires current + new; re-hashes and invalidates other sessions).</summary>
    public async Task<BoolResult> ChangePasswordAsync(string currentPassword, string newPassword, CancellationToken ct = default)
    {
        var body = new { currentPassword, newPassword };
        return await _http.PostAsync<BoolResult>("/api/v1/private/auth/change-password", body, ct).ConfigureAwait(false);
    }

    /// <summary>Create a short-lived handoff token for mobile app deep-link auth.</summary>
    public async Task<JsonElement> MobileHandoffAsync(CancellationToken ct = default)
    {
        return await _http.PostAsync<JsonElement>("/api/v1/private/auth/mobile-handoff", null, ct).ConfigureAwait(false);
    }

    /// <summary>Return caller's permission flags (capabilities, tier, feature gates).</summary>
    public async Task<JsonElement> PermissionsAsync(CancellationToken ct = default)
    {
        return await _http.GetAsync<JsonElement>("/api/v1/private/auth/permissions", ct: ct).ConfigureAwait(false);
    }

    /// <summary>Rotate session JWT (sliding expiry); returns new token.</summary>
    public async Task<LoginResult> RefreshAsync(CancellationToken ct = default)
    {
        return await _http.PostAsync<LoginResult>("/api/v1/private/auth/refresh", null, ct).ConfigureAwait(false);
    }
}

/// <summary>
/// MFA API — enroll and manage TOTP multi-factor authentication.
/// Maps to <c>/api/v1/private/mfa/*</c>.
/// </summary>
public sealed class MfaApi
{
    private readonly MelayaHttpClient _http;

    internal MfaApi(MelayaHttpClient http) => _http = http;

    /// <summary>Return MFA enrollment status for the caller.</summary>
    public async Task<MfaStatus> StatusAsync(CancellationToken ct = default)
    {
        return await _http.GetAsync<MfaStatus>("/api/v1/private/mfa/status", ct: ct).ConfigureAwait(false);
    }

    /// <summary>Initiate TOTP setup; returns QR code data and TOTP secret.</summary>
    public async Task<MfaSetupResult> SetupAsync(CancellationToken ct = default)
    {
        return await _http.PostAsync<MfaSetupResult>("/api/v1/private/mfa/setup", null, ct).ConfigureAwait(false);
    }

    /// <summary>Confirm TOTP setup with first valid code; activates MFA.</summary>
    public async Task<BoolResult> ConfirmSetupAsync(string code, CancellationToken ct = default)
    {
        var body = new { code };
        return await _http.PostAsync<BoolResult>("/api/v1/private/mfa/confirm", body, ct).ConfigureAwait(false);
    }
}

File: packages/sdk-php/src/AuthAPI.php
<?php

declare(strict_types=1);

namespace Melaya;

/**
 * Auth API — login, register, MFA, password management, session control.
 *
 * Maps to /api/v1/private/auth/* and /api/v1/auth/* (public endpoints).
 *
 * @example
 * ```php
 * $sdk = new Melaya\Melaya(apiKey: getenv('MK'));
 * $me  = $sdk->auth->me();
 * echo $me['username'];
 * ```
 */
class AuthAPI
{
    public function __construct(private readonly HttpClient $http) {}

    // ── Public (no auth required) ────────────────────────────────────────────

    /**
     * Login with username + password.
     * Returns a session JWT (and optional MFA challenge token if MFA is enrolled).
     */
    public function login(string $username, string $password): array
    {
        return $this->http->post('/api/v1/private/auth/login', [
            'username' => $username,
            'password' => $password,
        ]);
    }

    /**
     * Resolve an MFA challenge after login.
     * Exchange challenge token + TOTP code (or recovery code) for a full session JWT.
     */
    public function verifyMfa(string $challengeToken, string $code): array
    {
        return $this->http->post('/api/v1/private/auth/mfa/verify', [
            'challengeToken' => $challengeToken,
            'code'           => $code,
        ]);
    }

    /**
     * Register a new user account.
     * Sends an email verification link.
     */
    public function register(string $username, string $email, string $password): array
    {
        return $this->http->post('/api/v1/private/auth/register', [
            'username' => $username,
            'email'    => $email,
            'password' => $password,
        ]);
    }

    /** Confirm email address from the signup verification link token. */
    public function verifySignup(string $token): array
    {
        return $this->http->post('/api/v1/private/auth/verify-signup', ['token' => $token]);
    }

    /** Re-send the email verification message. */
    public function resendVerification(string $email): array
    {
        return $this->http->post('/api/v1/private/auth/resend-verification', ['email' => $email]);
    }

    /** Initiate a password reset flow; sends email with reset token. */
    public function forgotPassword(string $email): array
    {
        return $this->http->post('/api/v1/private/auth/forgot-password', ['email' => $email]);
    }

    /** Complete password reset using the token from the reset email. */
    public function resetPassword(string $token, string $newPassword): array
    {
        return $this->http->post('/api/v1/private/auth/reset-password', [
            'token'       => $token,
            'newPassword' => $newPassword,
        ]);
    }

    // ── Authenticated ────────────────────────────────────────────────────────

    /**
     * Return the current authenticated user profile.
     * Fields: id, username, email, tier, role, capabilities.
     */
    public function me(): array
    {
        return $this->http->get('/api/v1/private/auth/me');
    }

    /** Lightweight session validity check. Returns ['ok' => true] if valid. */
    public function check(): array
    {
        return $this->http->get('/api/v1/private/auth/check');
    }

    /** Change password (requires current password). Invalidates other sessions. */
    public function changePassword(string $currentPassword, string $newPassword): array
    {
        return $this->http->post('/api/v1/private/auth/change-password', [
            'currentPassword' => $currentPassword,
            'newPassword'     => $newPassword,
        ]);
    }

    /** Create a short-lived handoff token for mobile app deep-link auth. */
    public function createMobileHandoff(): array
    {
        return $this->http->post('/api/v1/private/auth/mobile-handoff');
    }

    /** Return the caller's permission flags (capabilities, tier, feature gates). */
    public function permissions(): array
    {
        return $this->http->get('/api/v1/private/auth/permissions');
    }

    /** Rotate the session JWT (sliding expiry). Returns a new token. */
    public function refresh(): array
    {
        return $this->http->post('/api/v1/private/auth/refresh');
    }

    // ── MFA management ───────────────────────────────────────────────────────

    /** Return MFA enrollment status for the caller. */
    public function mfaStatus(): array
    {
        return $this->http->get('/api/v1/private/mfa/status');
    }

    /** Initiate TOTP setup. Returns a QR code URL and secret. */
    public function mfaSetup(): array
    {
        return $this->http->post('/api/v1/private/mfa/setup');
    }

    /** Confirm TOTP setup with the first valid code. Activates MFA. */
    public function mfaConfirm(string $code): array
    {
        return $this->http->post('/api/v1/private/mfa/confirm', ['code' => $code]);
    }

    // ── Version ──────────────────────────────────────────────────────────────

    /** Get the current server version string (public). */
    public function version(): array
    {
        return $this->http->get('/api/v1/version');
    }
}
